### ISSUE TITLE
Callbacks not being called when app starts from a notification

### ISSUE THREAD

**Issue by andiradulescu** [OneSignal/OneSignal-iOS-SDK#115]
Although I see a UIApplicationLaunchOptionsRemoteNotificationKey in launchOptions in initWithLaunchOptions, the callbacks are not being called.

After taking a quick look inside the OneSignal codebase, I don't see where the notification payload is treated. But keep in mind that I'm not familiar with the code, since I started using OneSignal for less than a day.

Please let me know if there is a issue with the current version of the SDK.

Thanks!

**Comment by jkasten2** [comment]
@andiradulescu The notification payload is being process on the following line.
https://github.com/OneSignal/OneSignal-iOS-SDK/blob/2.1.16/iOS_SDK/OneSignal/OneSignal.m#L1392

Do you have any other push notifications SDKs in your app? Are you calling any of Apple's notification functions? If so can you try temporarily disabling them? 

Thanks.

**Comment by andiradulescu** [comment]
Just OneSignal, and I don't seem to call any of the Apple's notification functions. I will test the functionality on the sample project and get back.

Thanks

**Comment by andiradulescu** [comment]
@jkasten2 I finally found the issue.

I tested with both master (=2.2.0) and 2.1.16.

The thing is that the app was calling OneSignal.initWithLaunchOptions later in the loading process, after didFinishLaunchingWithOptions finished. After looking through the code and seeing the swizzling, I had a hunch that the swizzle happened after the Apple notification function was called. So I moved the init inside didFinishLaunchingWithOptions and everything went well.

You're probably thinking why OneSignal.init is being called later in the execution of the app. It's being called later because the option to enable push notifications is coming in a config response from the server. So the app waits everytime to see if push notifications are enabled in order to call OneSignal.init.

How can this situation be resolved?

Anyway, on 2.2.0 it even crashes in this case. Just call OneSignal.initWithLaunchOptions somewhere else in the app, after didFinishLaunchingWithOptions returned.

**Comment by jkasten2** [comment]
@andiradulescu Thanks for the details, you can set [`kOSSettingsKeyAutoPrompt`](https://documentation.onesignal.com/docs/ios-native-sdk#section--kossettingskeyautoprompt-) to `false` to the `NSDictionary` `settings:` passed to [`initWithLaunchOptions`](https://documentation.onesignal.com/docs/ios-native-sdk#section--initwithlaunchoptions-) to keep our SDK from prompting.

Could you attach the full stack trace you are seeing in 2.2.0 with the late call to `initWithLaunchOptions`?

Thanks.

**Comment by andiradulescu** [comment]
@jkasten2 did as you said and is working fine! Thanks!

I will attach the stack trace, didn't had the inspiration to save a screenshot when I saw it.

If I don't respond in 24h, please remind me.

**Comment by andiradulescu** [comment]
@jkasten2 here is the screenshot of the crash: http://snpy.in/CjnZSU

I also changed the sample to crash with a "dispatch after": https://www.dropbox.com/s/r25ll0z2z74s4yh/OneSignal-iOS-SDK-master.zip?dl=0

**Comment by jkasten2** [comment]
@andiradulescu Thanks again for the report and details. I was able to reproduce the crash with your delay code when opening a notifications after force killing the app. It was fixed in commit f0d52533a0dd7b7462c2d31f6ea2a291066f87b1 and will be included in the next release. You will still need to ensure however that OneSignal `initWithLaunchOptions ` is called from `didFinishLaunchingWithOptions` for all the events to work correctly.
